pkg/controller/replica.go

```diff
@@ -0,0 +1,329 @@
+	// this is to prevent multiple sync volume operation to be performed on a single volume concurrently as it can create or delete more attachments than necessary
+	v, _ := r.mutexLocks.LoadOrStore(volumeName, &sync.RWMutex{})
+	volumeLock := v.(*sync.RWMutex)
+	volumeLock.Lock()
+	defer volumeLock.Unlock()
+
+	azVolume, err := azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volumeName, r.namespace, true)
+	if errors.IsNotFound(err) {
+		klog.Infof("Aborting replica management... volume (%s) does not exist", volumeName)
+		return nil
+	} else if err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+	if !isCreated(azVolume) {
+		return status.Errorf(codes.Aborted, "azVolume (%s) has no underlying volume object", azVolume.Name)
+	}
+
+	// get all replica attachments for the given volume
+	volReq, err := createLabelRequirements(azureutils.VolumeNameLabel, volumeName)
+	if err != nil {
+		return err
+	}
+	roleReq, err := createLabelRequirements(azureutils.RoleLabel, string(v1alpha1.ReplicaRole))
+	if err != nil {
+		return err
+	}
+	labelSelector := labels.NewSelector().Add(*volReq, *roleReq)
+	azVolumeAttachments := &v1alpha1.AzVolumeAttachmentList{}
+	err = r.client.List(ctx, azVolumeAttachments, &client.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		klog.Errorf("failed to list AzVolumeAttachment: %v", err)
+		return err
+	}
+
+	desiredReplicaCount, currentReplicaCount := azVolume.Spec.MaxMountReplicaCount, len(azVolumeAttachments.Items)
+	klog.Infof("control number of replicas for volume (%s): desired=%d,\tcurrent:%d", azVolume.Spec.UnderlyingVolume, desiredReplicaCount, currentReplicaCount)
+
+	// if the azVolume is marked deleted, do no create more azvolumeattachment objects
+	if azVolume.DeletionTimestamp == nil && desiredReplicaCount > currentReplicaCount {
+		klog.Infof("Need %d more replicas for volume (%s)", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume)
+		if azVolume.Status.Detail == nil || azVolume.Status.State == v1alpha1.VolumeDeleting || azVolume.Status.State == v1alpha1.VolumeDeleted || azVolume.Status.Detail.ResponseObject == nil {
+			// underlying volume does not exist, so volume attachment cannot be made
+			return nil
+		}
+		if err = r.createReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredReplicaCount-currentReplicaCount), azVolume.Name, azVolume.Status.Detail.ResponseObject.VolumeID); err != nil {
+			klog.Errorf("failed to create %d replicas for volume (%s): %v", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileReplica) getNodesForReplica(ctx context.Context, volumeName string, numReplica int) ([]string, error) {
+	pods, err := r.controllerSharedState.getPodsFromVolume(volumeName)
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range pods {
+		v, _ := r.podLocks.LoadOrStore(pod, &sync.RWMutex{})
+		podLock := v.(*sync.RWMutex)
```

Comment:
Why are we acquiring a lock here when the `SharedState` we're using can be update by other controllers? This method also appears to only ready state, so why an exclusive lock? (For that matter, if only the enxclusive lock is used, why a `RWMutex`?)

Reply:
The purpose of this lock is to ensure that replica management happens serially for each pod; this is to ensure that replica management for volumes that share a pod don't run into concurrency issue (e.g. if `ManageReplica` is called concurrently for volume 1 and 2 sharing a pod, they might end up being placed on different nodes).

That said, I am wondering if pod controller and replica controller should be sharing this lock as I see scenarios where we can have concurrency issue between the two controllers as well.

I agree, `RWMutex` is not needed here; I will convert it to `Mutex`

Reply:
I don't see where in this function that replicas are being managed, thus the confusion. This method is called by a function which uses the result to create replicas, but the lock is released on returned and not held while the replicas are created.

Reply:
You are right. I was completely mistaken that the replicas were being created within this function. I need to rework this part. Thanks for pointing this out